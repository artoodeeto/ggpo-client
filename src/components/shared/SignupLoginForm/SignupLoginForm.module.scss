@import '../../../assets/media-queries.scss';
@import '../../../assets/color-theme-dark.scss';

.SignUpLoginForm {
  display: flex;
  flex-direction: column;
  align-items: center;

  > input {
    box-sizing: border-box;
    width: 100%;
    padding: 12px;
    font-size: 14px;
    margin: 0.2em;
    border-radius: 4px;
    border: 0px;
  }

  > input[type='submit'] {
    margin-top: 5px;
    background-color: $blue-cheese;
    font-weight: bolder;
    font-size: 0.8em;
    color: $ashy-foot;
  }

  > input[type='email']:not(:placeholder-shown):invalid {
    color: red;
  }
}
