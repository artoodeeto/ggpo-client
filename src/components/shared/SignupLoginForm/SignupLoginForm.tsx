import React, { FC, FormEvent, useState } from 'react';
import { ILoginSignUpFormParams } from 'interfaces/session';
import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { RootState } from 'store/root/root_reducer';
import { logMeIn, signMeUp } from 'store/session/Actions';
import styles from './SignupLoginForm.module.scss';

const handleSubmit = (
  event: FormEvent,
  params: ILoginSignUpFormParams,
  signupOrLoginAPI: (userSignup: ILoginSignUpFormParams) => void
): void => {
  event.preventDefault();

  signupOrLoginAPI(params);
};

type SignupLoginFormProps = {
  toLoginOrSignup: boolean;
  onLogin: (userLogin: ILoginSignUpFormParams) => void;
  onSignup: (userSignup: ILoginSignUpFormParams) => void;
};

// TODO: change this as array of objects so we wont have render each input. we'll replace this with filter.
/**
 * TODO: sample object
 * [{
 *  onchange: fn,
 *  value,
 *  placeHolder,
 * type
 * }]
 */
export const SignupLoginForm: FC<SignupLoginFormProps> = ({ toLoginOrSignup, onLogin, onSignup }) => {
  const [email, setEmail] = useState('');
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const params = {
    email,
    username,
    password
  };

  return (
    <form
      className={styles.SignUpLoginForm}
      onSubmit={(e) => handleSubmit(e, params, toLoginOrSignup ? onLogin : onSignup)}
    >
      <input
        onChange={(e) => setEmail(e.target.value)}
        value={email}
        placeholder="email_address@gmail.com"
        type="text"
        name="email"
        id=""
        required
      />

      {toLoginOrSignup ? null : (
        <input
          onChange={(e) => setUsername(e.target.value)}
          value={username}
          placeholder="user name"
          type="text"
          name="username"
          id=""
          required
        />
      )}

      <input
        onChange={(e) => setPassword(e.target.value)}
        value={password}
        placeholder="password"
        type="text"
        name="password"
        id=""
        required
      />

      {toLoginOrSignup ? null : (
        <input
          onChange={(e) => setConfirmPassword(e.target.value)}
          value={confirmPassword}
          placeholder="confirm password"
          type="text"
          name="conf_password"
          id=""
          required
        />
      )}

      <input type="submit" value={toLoginOrSignup ? 'Login' : 'Signup'} />
    </form>
  );
};

const mapStateToProps = (state: RootState) => {
  return {};
};

const mapDispatchToProps = (dispatch: ThunkDispatch<RootState, {}, AnyAction>) => {
  return {
    onSignup: (userSignup: ILoginSignUpFormParams) => dispatch(signMeUp(userSignup)),
    onLogin: (userLogin: ILoginSignUpFormParams) => dispatch(logMeIn(userLogin))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(SignupLoginForm);
