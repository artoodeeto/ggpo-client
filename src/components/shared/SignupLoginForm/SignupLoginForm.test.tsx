/* eslint-disable @typescript-eslint/no-empty-function */
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import { SignupLoginForm } from './SignupLoginForm';

describe('<SignupLoginForm/>', () => {
  describe('snapshot', () => {
    it('User loging in snapshot', () => {
      const wrapper = shallow(<SignupLoginForm toLoginOrSignup onLogin={() => {}} onSignup={() => {}} />);
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('User signing up snapshot', () => {
      const wrapper = shallow(<SignupLoginForm toLoginOrSignup={false} onLogin={() => {}} onSignup={() => {}} />);
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('form tag', () => {
    it('form should have class name of SignUpLoginForm', () => {
      const wrapper = shallow(<SignupLoginForm toLoginOrSignup={false} onLogin={() => {}} onSignup={() => {}} />);
      expect(wrapper.find('form')).toHaveClassName('SignUpLoginForm');
    });
  });

  describe('User signing up', () => {
    const wrapper = shallow(<SignupLoginForm toLoginOrSignup={false} onLogin={() => {}} onSignup={() => {}} />);
    const inputs = wrapper.find('input');
    it('should display 5 input tags', () => {
      expect(inputs.length).toEqual(5);
    });
    it('input type=button should have value of Signup', () => {
      expect(inputs.at(4).props().value).toBe('Signup');
    });
  });

  describe('User logging in', () => {
    const wrapper = shallow(<SignupLoginForm toLoginOrSignup onLogin={() => {}} onSignup={() => {}} />);
    const inputs = wrapper.find('input');
    it('should display 5 input tags', () => {
      expect(inputs.length).toEqual(3);
    });
    it('input type=button should have value of Login', () => {
      expect(inputs.at(2).props().value).toBe('Login');
    });
  });
});
