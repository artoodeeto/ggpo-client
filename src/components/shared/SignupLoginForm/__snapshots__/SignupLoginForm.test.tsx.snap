// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupLoginForm/> snapshot User loging in snapshot 1`] = `
<form
  className="SignUpLoginForm"
  onSubmit={[Function]}
>
  <input
    id=""
    name="email"
    onChange={[Function]}
    placeholder="email"
    required={true}
    type="email"
    value=""
  />
  <input
    aria-describedby="password-constraints"
    id=""
    name="password"
    onChange={[Function]}
    placeholder="password"
    required={true}
    type="text"
    value=""
  />
  <input
    type="submit"
    value="Login"
  />
</form>
`;

exports[`<SignupLoginForm/> snapshot User signing up snapshot 1`] = `
<form
  className="SignUpLoginForm"
  onSubmit={[Function]}
>
  <input
    id=""
    name="email"
    onChange={[Function]}
    placeholder="email"
    required={true}
    type="email"
    value=""
  />
  <input
    id=""
    name="username"
    onChange={[Function]}
    placeholder="user name"
    required={true}
    type="text"
    value=""
  />
  <input
    aria-describedby="password-constraints"
    id=""
    name="password"
    onChange={[Function]}
    placeholder="password"
    required={true}
    type="text"
    value=""
  />
  <input
    id=""
    name="conf_password"
    onChange={[Function]}
    placeholder="confirm password"
    required={true}
    type="text"
    value=""
  />
  <input
    type="submit"
    value="Signup"
  />
</form>
`;
